expression/builtin_math.go

```diff
@@ -1743,6 +1743,10 @@ func (c *truncateFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 	if argTp == types.ETTimestamp || argTp == types.ETDatetime || argTp == types.ETDuration || argTp == types.ETString {
 		argTp = types.ETReal
 	}
+	// MySQL compatibility
+	if argTp == types.ETJson {
+		argTp = types.ETInt
+	}
```

Comment:
I was really curious why the Type is `ETInt` instead of `ETReal`. And I run the SQL in my local MySQL 5.7.26 and MySQL 8.0.17. It gets an interesting result.
```
mysql> select truncate(json_array(), 1), truncate("cascasc", -1), truncate("12.3333", -1);
+---------------------------+-------------------------+-------------------------+
| truncate(json_array(), 1) | truncate("cascasc", -1) | truncate("12.3333", -1) |
+---------------------------+-------------------------+-------------------------+
|                       0.0 |                       0 |                      10 |
+---------------------------+-------------------------+-------------------------+
1 row in set, 2 warnings (0.00 sec)

mysql> show warnings;
+---------+------+-----------------------------------------------------------------------+
| Level   | Code | Message                                                               |
+---------+------+-----------------------------------------------------------------------+
| Warning | 3156 | Invalid JSON value for CAST to DOUBLE from column json_array at row 1 |
| Warning | 1292 | Truncated incorrect DOUBLE value: 'cascasc'                           |
+---------+------+-----------------------------------------------------------------------+
2 rows in set (0.00 sec)
```
And it seems that the function cast json to double ?

Reply:
Fixed. For all types with `IsStringKind`, returns real type.